### PR TITLE
Add wrapper type for safety

### PR DESCRIPTION
**Pull request by jstarry**
#### Summary of Changes

Fixes #

### CODE REVIEW

File: runtime/src/builtins.rs [hunk 1]
```diff
@@ -141,7 +150,14 @@ pub enum BuiltinFeatureTransition {
 unsafe impl Send for BuiltinFeatureTransition {}
 unsafe impl Sync for BuiltinFeatureTransition {}
 
-impl BuiltinFeatureTransition {
+// The following trait bounds ensure that the above unsafe trait implementations
+// won't allow undefined behavior to be introduced.
+#[cfg(debug_assertions)]
+trait AutoSendSync: Send + Sync {}
```
Comment (t-nelson) [inline runtime/src/builtins.rs:156]: Sorry how does this work?  My understanding of how marker traits are handled is that they just tell the auto-trait resolver, "stop digging, you've struck gold."  There's no compiler side verification of `Send`-iness of `Sync`-iness, which is why they're `unsafe`.

Comment (jstarry) [reply runtime/src/builtins.rs:156]: A type can't implement AutoSendSync unless it satisfies the Send + Sync trait requirements as well

Comment (t-nelson) [reply runtime/src/builtins.rs:156]: Yeah I understand that, but it doesn't prevent UB.  I can still implement `AutoSendSync` on a type that I've screwed up `Send` and/or `Sync` implementations on.  So I guess the comment confused me at best.  I'm not sure I understand the value add here other than maybe to mark any other types we may hit this bug on before Rust 1.60.0?

To be clear, the issue the original PR is fixing is not UB, it's bypassing a bug in the auto-trait resolver

Comment (jstarry) [reply runtime/src/builtins.rs:156]: You can't implement AutoSendSync unless a type already implements Send + Sync. So if you break the type, the AutoSendSync will cause a compiler error. Without it, your unsafe impl could allow new UB to be undetected

Comment (t-nelson) [reply runtime/src/builtins.rs:156]: Ok I get it after some offline brain bashing.  Didn't realize that the trait resolver would check field types on explicitly declared marker traits.

I'm cool with this.  Going to move it to sdk and mark it deprecated under the assumption we'll need it again

**Comment by t-nelson** [comment]
Closing in favor of https://github.com/solana-labs/solana/pull/23321

**Pull request closed without merging**